assumeutxo: add init and completion logic

jamesob opened the pull request:
This is part of the [assumeutxo project](https://github.com/bitcoin/bitcoin/projects/11) (parent PR: #15606) 

---

This changeset adds logic for detecting and activating snapshot chainstates on start, and for completing the snapshot validation process once the background chainstate reaches the base block of the assumed-valid chain. It also handles removing the snapshot chainstate data on-disk in the event that the snapshot fails to validate.

As detailed in [assumeutxo.md](https://github.com/bitcoin/bitcoin/blob/master/doc/assumeutxo.md#background-chainstate-hits-snapshot-base-block):

> Once the tip of the background chainstate hits the base block of the snapshot chainstate, we stop use of the background chainstate by setting m_stop_use, in CompleteSnapshotValidation(), which is checked in ActivateBestChain()). We hash the background chainstate's UTXO set contents and ensure it matches the compiled value in CMainParams::m_assumeutxo_data.
> 
> The background chainstate data lingers on disk until shutdown, when in ChainstateManager::Reset(), the background chainstate is cleaned up with ValidatedSnapshotShutdownCleanup(), which renames the chainstate_[hash] datadir as chainstate.
>
> **Failure consideration**: if bitcoind unexpectedly halts after m_stop_use is set on the background chainstate but before CompleteSnapshotValidation() can finish, the need to complete snapshot validation will be detected on subsequent init by ChainstateManager::CheckForUncleanShutdown().

---

~~Most of this change is unittested, though some logic (anything assuming on-disk leveldb data) can't be tested without some rework of unittest setup, given in-memory leveldb instances are used. I can look into how difficult this would be.~~

This change adds the ability to unittest using on-disk leveldb coins dbs.

#### Possible follow-ups

- [ ] Unify `CDBWrapper::{m_path,m_is_memory}`: https://github.com/bitcoin/bitcoin/pull/24232#discussion_r855354180

jamesob committed:
validation: remove snapshot datadirs upon validation failure

If a UTXO snapshot fails to validate, don't leave the resulting datadir
on disk as this will confuse initialization on next startup and we'll
get an assertion error.

jamesob committed:
add ChainstateManager.GetSnapshot{BaseHeight,BaseBlock}()

For use in later commits.

jamesob committed:
db: add StoragePath to CDBWrapper/CCoinsViewDB

This is used in subsequent commits. It allows us to clean up UTXO
snapshot chainstate after background validation completes.

MarcoFalke commented on src/validation.h:
```diff
@@ -977,12 +1014,27 @@ class ChainstateManager
 
+    //! When starting up, search the datadir for a chainstate based on a UTXO
+    //! snapshot that is in the process of being validated.
+    bool DetectSnapshotChainstate(CTxMemPool& mempool) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
any reason this is a reference. `memepool` at the call site is already a pointer to allow setting it to nullptr

MarcoFalke commented on src/validation.h:
```diff
@@ -977,12 +1014,27 @@ class ChainstateManager
+    //! background validating and has reached the base block of the snapshot.
+    bool MaybeCompleteSnapshotValidation(
+        CChainState& chainstate,
+        CBlockIndex* pindexNew) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
Any reason this is a pointer? This can't be null, neither at the call site nor inside the function.

MarcoFalke commented on src/node/chainstate.cpp:
```diff
@@ -28,10 +28,19 @@ std::optional<ChainstateLoadingError> LoadChainstate(bool fReset,
+    // create a validation chainstate.
+    if (!chainman.IsSnapshotValidated()) {
+        LogPrintf("Loading validation chainstate\n");
+        chainman.InitializeChainstate(Assert(mempool));
```
any reason to add the assert?

maflcko reviewed (commented):
Looks like there are test failures

jamesob committed:
blockmanager: avoid undefined behavior during FlushBlockFile

If we call FlushBlockFile() without having intitialized the block index
with LoadBlockIndexDB(), we may be indexing into an empty vector.

Specifically this is an issue when we call MaybeRebalanceCaches() during
chainstate init before the block index has been loaded, which calls
FlushBlockFile().

Also add an assert to avoid undefined behavior.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>

jamesob commented:
Thanks for the quick feedback Mco. 

The test failure was actually due to removing a commit that I'd forgotten the use for. Basically, because we're now calling MaybeRebalanceCaches in LoadChainstate, that prompts a FlushStateToDisk call before the block index is loaded, which trips some UB. 

The bounds check in https://github.com/bitcoin/bitcoin/pull/24232/commits/d70b8de51e54061e9ef045366f4a7651678dd82a fixes this.

jamesob committed:
validation: add CChainState::m_stop_use and ChainMan::isUsable

and remove m_snapshot_validated. This state can now be inferred by the
number of isUsable chainstates.

m_stop_use is used to signal that a chainstate should no longer be used
by validation logic; it is used as a sentinel when background validation
completes or if the snapshot chainstate is found to be invalid.

isUsable is a convenience method that incorporates m_stop_use.

jamesob committed:
add CChainState::HasCoinsViews()

Used in subsequent commits. Also cleans up asserts in
coins_views-related convenience methods to be more exact.

jamesob commented on src/validation.cpp:
```diff
@@ -4982,6 +4991,91 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+
+    if (!GetUTXOStats(&ibd_coins_db, WITH_LOCK(::cs_main, return std::ref(m_blockman)), ibd_stats, breakpoint_fnc)) {
+        LogPrintf("[snapshot] failed to generate stats for validation coins db\n");
+        return false;
```
Should probably de-activate the existing snapshot chainstate as in the `if (snapshot_invalid)` conditional.

jamesob committed:
validation: add ChainMan logic for completing UTXO snapshot validation

Trigger completion when a background validation chainstate reaches the
same height as a UTXO snapshot, and handle cleaning up the chainstate on
subsequent startup.

jamesob committed:
move-only: test: make snapshot chainstate setup reusable

For use in next commit.

Most easily reviewed with
`--color-moved=dimmed_zebra --color-moved-ws=allow-indentation-change`.

jamesob committed:
test: add testcases for snapshot completion

jamesob commented:
Okay, I've pushed a revision that addresses my previous comments and more comprehensively (that is to say: at all) tests the major failure modes. I had to parameterize the shutdown behavior during snapshot completion to get this to work within the unittests.

Pending CI pass, this should be ready for continued review.

jamesob committed:
test: add reset_chainstate parameter for snapshot unittests

This CreateAndActivateUTXOSnapshot parameter is necessary once we
perform snapshot completion within ABC, since the existing UpdateTip
test will fail because the IBD chain that has generated the snapshot
will exceed the base of the snapshot.

Being able to test snapshots being loaded into a mostly-uninitialized
datadir allows for more realistic unittest scenarios.

jamesob commented:
Alright, CI is finally passing. I had to add https://github.com/bitcoin/bitcoin/pull/24232/commits/5d5dfaa7a647307f58acdd1909fc8f34c1787f82 because an existing unittest that advanced an IBD chain past the snapshot it generated started failing. This isn't something that should happen in practice, but was done to test an unrelated feature (`UpdateTip` behavior on a background chainstate).

jonatack commented on src/validation.cpp:
```diff
@@ -5009,26 +5129,54 @@ void ChainstateManager::Unload()
 void ChainstateManager::Reset()
 {
-    LOCK(::cs_main);
+    auto get_storage_path = [](auto& chainstate) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) -> std::optional<fs::path> {
```
This annotation in the definition is redundant, as it is already in the declaration.

MarcoFalke commented on src/test/util/chainstate.h:
```diff
@@ -48,6 +60,33 @@ CreateAndActivateUTXOSnapshot(node::NodeContext& node, const fs::path root, F ma
+            // removing. Without those headers, we can't activate the snapshot below.
+            LOCK(::cs_main);
+            uint256 gen_hash = node.chainman->ActiveChainstate().m_chain[0]->GetBlockHash();
+            node.chainman->Reset();
```
Looks like #24299 was merged, so you might have to inline the needed resets here (or create a helper function for it).

jamesob commented:
Alright, have rebased; basically I just took the old `ChainstateManager::Reset()` method and renamed it to `cleanup()`, and applied @jonatack -style changes (lock assertion/annotation vs. LOCK).

jonatack commented:
Planning to review soon.

ryanofsky commented on src/validation.h:
```diff
@@ -492,6 +492,12 @@ class CChainState
+    //! This toggle exists for use when doing background validation for UTXO
+    //! snapshots. It is set once the background validation chain reaches the
+    //! same height as the base of the snapshot, and signals that we should no
+    //! longer connect blocks to this chainstate.
```
In commit "validation: add CChainState::m_stop_use and ChainMan::isUsable" (c703836160f3eca8073c22a1531bb98ad9666999)

Would `s/this chainstate/the background chainstate/`. Which chainstate is "this chainstate" was not immediately obvious.

jamesob replied:
Fixed

ryanofsky commented on src/validation.h:
```diff
@@ -866,6 +872,15 @@ class ChainstateManager
 
+    //! Return true if a chainstate is considered usable.
+    //!
+    //! This might be false for, say, a background validation chainstate which has
```
In commit "validation: add CChainState::m_stop_use and ChainMan::isUsable" (c703836160f3eca8073c22a1531bb98ad9666999)

Would change "This might be false, for, say" to "This will be false for". IMO wishy washy comments are hard to understand and don't provide any real future proofing (A veritably false comment is still less confusing than a technically true but mostly misleading comment)

jamesob replied:
Wishywashy comment no more.

ryanofsky reviewed (commented):
Started review (will update list below with progress).
- [X] becdbd6da4c6d7c9c658ffc38b1c9da1c89571a1 init: add utxo snapshot detection (1/11)
- [X] fe09582ec34ff4d19bb2aa490ffd6ccca95e8480 validation: tear down snapshot datadirs upon validation failure (2/11)
- [X] d8f32839cfbc8cbfe194d190bdc19fb3c4d6e32a add ChainstateManager.getSnapshot{Height,BaseBlock}() (3/11)
- [X] b28ed6628850f2f4eceb33c059c0d9997972bbc0 db: add StoragePath() to CDBWrapper and CCoinsViewDB (4/11)
- [X] 1aeca9ae4d9938a9ae5a06dbe16d8b1cd40b9960 blockmanager: avoid undefined behavior during FlushBlockFile (5/11)
- [X] c703836160f3eca8073c22a1531bb98ad9666999 validation: add CChainState::m_stop_use and ChainMan::isUsable (6/11)
- [X] bf36616fc62f668c45d30449fb5b7c2c84a73a99 add CChainState::HasCoinsViews() (7/11)
- [x] 328aa05c28fe9e5fff5d90e9b23bbbb5f6da1846 validation: add ChainMan logic for completing UTXO snapshot validation (8/11)
- [x] b1c55101c2883bf03777c320cec3b05b58454a53 move-only: test: make snapshot chainstate setup reusable (9/11)
- [x] 9916bb825f6ef69f6c906dd4154309df07aa9d03 test: add reset_chainstate parameter for snapshot unittests (10/11)
- [ ] 51d0dd34ecfb42d80707ec6cf788f0e63e30439f test: add testcases for snapshot completion (11/11)

ryanofsky commented on src/validation.h:
```diff
@@ -862,6 +862,10 @@ class ChainstateManager
 
+    // Returns nullptr if no snapshot ahs been loaded.
+    CBlockIndex* getSnapshotBaseBlock() EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+    std::optional<int> getSnapshotHeight() EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
In commit "add ChainstateManager.getSnapshot{Height,BaseBlock}()" (d8f32839cfbc8cbfe194d190bdc19fb3c4d6e32a)

Would rename `getSnapshotHeight` to `getSnapshotBaseHeight` for two reasons:

- To be consistent with `getSnapshotBaseBlock`
- To disambiguate. "Snapshot" could refer to snapshot chainstate rather than snapshot itself, and chainstate may have a height higher than the initial snapshot

This is a tangent, but I'm increasingly disliking the terms "ibd chainstate" and "snapshot chainstate." If had a magic wand I would abolish them everywhere. I think ideally a chainstate would consist of a UTXO set, a tip pointer, and "validated starting from" block pointer and "validated starting from" UTXO hash. ChainstateManager would have a list of chainstates, and one of them would be active with rest inactive. Network code would download blocks to validate chainstates, prioritizing blocks for the active chainstate. ChainstateManager would connect the incoming blocks, and if the tip of any chainstate crossed the "validated starting from" block of a later chainstate, and UTXO hash matched, the later "validated_starting_from" pointer would moved back to the earlier one, and earlier chainstate (optionally) discarded.
Probably having explicit tip and "validated starting from" pointers would also allow CBlockIndex validation flags to be simplified so more code doesn't need to care about assumeutxo. This could all be cleanup though. More important to get the existing code that's already merged functional and not change things at this point.

jamesob replied:
Renamed. Agree with your thoughts on the better design, and also agree that the path from here to there is not so long. I'd just prefer to wait until we actually need to revamp the design (e.g. multiple parallel background chainstates) to do all the code changes.

ryanofsky commented on src/validation.cpp:
```diff
@@ -5051,6 +5172,34 @@ void ChainstateManager::Unload()
     m_best_invalid = nullptr;
 }
 
+void ChainstateManager::cleanup()
```
In commit "validation: add ChainMan logic for completing UTXO snapshot validation" (328aa05c28fe9e5fff5d90e9b23bbbb5f6da1846)

I think I'd NACK the approach for trying to rename and delete things on shutdown. I don't see why it's not possible to rename and delete the snapshot when it is validated. I understand that maybe you want the pointers to be stable so can't delete the CChainstate object, but that doesn't seem like a reason why you can't close/destroy databases and rename directories.

If I'm missing something and it really is not possible to rename the chainstate directories while the node is running, the best time to rename them would be startup time, not shutdown time. Shutdown is the time to stop work, not begin it, and it's especially precarious if bitcoin is running on a constantly rebooting windows machine or as a systemd service, where the system will may nicely for a shutdown but quickly run out of patience and kill the process.

jamesob replied:
Good points, willfix.

jamesob replied:
I'm now doing the "cleanup" (i.e. deletion of background chainstate data) on startup, and not shutdown. I'd like to experiment with doing deletion inline in a follow-up PR.

ryanofsky reviewed (commented):
Now reviewed the full PR, but will wait for answers to some questions before adding ACK.

I do think this PR could be merged in its current form, but some things, particularly detecting chainstate directories on startup, and renaming snapshot chainstate directory when its validated could be handled in a more robust way. So would want to clean these things up or find out what future plans may be.

Overall this looks very good. Could be a candidate for review club, I think.

ariard commented on src/validation.cpp:
```diff
@@ -5082,3 +5083,48 @@ void ChainstateManager::MaybeRebalanceCaches()
+                found = it->path();
+                LogPrintf("[snapshot] found snapshot datadir %s\n", fs::PathToString(found));
+            } else {
+                LogPrintf("[snapshot] WARNING - detected multiple snapshot " /* Continued */
```
At least one improvement of the log message could be to advice the node operator on what they should do, e.g deleting the oldest snapshot one or whatever could be appropriate.

ariard commented on src/validation.cpp:
```diff
@@ -5128,3 +5130,28 @@ bool ChainstateManager::DetectSnapshotChainstate(CTxMemPool* mempool)
+    uint256 datadir_hash = PathToSnapshotHash(*snapshot_datadir);
+
+    if (datadir_hash != *m_from_snapshot_blockhash) {
+        LogPrintf("[snapshot] WARNING - unexpected blockhash for snapshot datadir (%s); expected %s\n",
```
Same comment, unclear what a node operator should do with an invalid, name-different snapshot datadir.

ariard commented on src/validation.h:
```diff
@@ -862,6 +862,10 @@ class ChainstateManager
         CBlockIndex** ppindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
     friend CChainState;
 
+    // Returns nullptr if no snapshot ahs been loaded.
```
nit: s/ahs/has/g

jamesob replied:
Fixed, thanks

ariard reviewed (commented):
Reviewed until bf36616

fanquake commented on src/test/validation_chainstate_tests.cpp:
```diff
@@ -90,7 +90,8 @@ BOOST_FIXTURE_TEST_CASE(chainstate_update_tip, TestChain100Setup)
 
-    BOOST_REQUIRE(CreateAndActivateUTXOSnapshot(m_node, m_path_root));
+    BOOST_REQUIRE(CreateAndActivateUTXOSnapshot(
+        this, NoMalleation, /* reset_chainstate */ true));
```
```suggestion
        this, NoMalleation, /*reset_chainstate=*/true));
```
Anywhere you are adding / touching named args, please use the [style from the developer docs](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c-named-arguments).

jamesob replied:
Done, thanks.

ariard reviewed (commented):
Reviewed 328aa05c, I think a commit can be added to update `bitcoin/assumeutxo.md` in consequence, the subsection "Background chainstate hits snapshots base block". Especially, the chainstate directories/caches dance done in function of completion success or failure.

jamesob commented on src/validation.cpp:
```diff
@@ -5082,3 +5083,48 @@ void ChainstateManager::MaybeRebalanceCaches()
+                found = it->path();
+                LogPrintf("[snapshot] found snapshot datadir %s\n", fs::PathToString(found));
+            } else {
+                LogPrintf("[snapshot] WARNING - detected multiple snapshot " /* Continued */
+                    "datadirs (%s). This is not expected.\n", fs::PathToString(it->path()));
```
> I think the best thing for bitcoin to do at startup would be to list all the chainstate* directories and use whichever one has the most work at its tip as the active chainstate.

I agree this would be a nice feature, but the process of unpacking snapshots and loading them to the point of being able to evaluate "most work" entails some big changes, so I'll defer this one for later. 

I do like your suggestion of calling it `<datadir>/chainstate_snapshot` and will incorporate that.

jamesob replied:
I've changed the code to write the snapshot chainstate leveldb to `chainstate_snapshot`. A separate file within that directory called `base_blockhash` contains a serialized uint256 of what used to be in the directory name.

jamesob commented on src/validation.cpp:
```diff
@@ -5025,6 +5035,117 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+        LogPrintf("[snapshot] !!! the snapshot you've been working off of is invalid\n");
+        LogPrintf("[snapshot] deleting snapshot and reverting to validated chain\n");
+
+        m_active_chainstate = m_ibd_chainstate.get();
```
`m_ibd_chainstate` is the active if a snapshot is not being used, so calling it `m_background_...` wouldn't be appropriate.

jamesob committed:
test: allow on-disk coins and block tree dbs in tests

Used when testing cleanup of on-disk chainstate data for snapshot
testcases. Also necessary for simulating node restart in .cpp tests.

jamesob commented on src/validation.cpp:
```diff
@@ -2959,6 +2964,12 @@ bool CChainState::ActivateBestChain(BlockValidationState& state, std::shared_ptr
 
         if (nStopAtHeight && pindexNewTip && pindexNewTip->nHeight >= nStopAtHeight) StartShutdown();
 
+        // This will have been toggled in ABCStep -> ConnectTip -> completeSnapshotValidation,
```
Fixed, thanks.

jamesob commented on src/validation.cpp:
```diff
@@ -5025,6 +5035,117 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+    }
+
+    auto handleInvalidSnapshot = [&]() EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
+        LogPrintf("[snapshot] !!! the snapshot you've been working off of is invalid\n");
+        LogPrintf("[snapshot] deleting snapshot and reverting to validated chain\n");
```
Now using `AbortNode()` as well as the nice log message you specify; I'm also including `PACKAGE_BUGREPORT` which lists the github issues URL. Thanks for the great suggestion.

jamesob commented on src/validation.cpp:
```diff
@@ -5025,6 +5035,117 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+         return {false, {}};
+     }
+
+    int curr_height = chainstate.m_chain.Height();
```
Done, thanks.

jamesob commented on src/validation.cpp:
```diff
@@ -5025,6 +5035,117 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+    CCoinsViewDB& ibd_coins_db = chainstate.CoinsDB();
+    chainstate.ForceFlushStateToDisk();
+
+    auto snapshot_blockhash = SnapshotBlockhash().value_or(uint256());
```
Done.

jamesob commented on src/validation.cpp:
```diff
@@ -5025,6 +5035,117 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+        chainstate.m_chain.Tip()->ToString(),
+        snapshot_blockhash.ToString());
+
+    assert(index_new.GetBlockHash() == snapshot_blockhash);
```
Good point, done.

jamesob commented on src/validation.cpp:
```diff
@@ -5025,6 +5035,117 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+        LogPrintf("[snapshot] failed to generate stats for validation coins db\n");
+        // While this isn't a proablem with the snapshot per se, this condition
+        // prevents us from validating the snapshot, so we should shut down and let the
+        // use handle the issue manually.
```
Done, thanks.

jamesob commented on src/validation.cpp:
```diff
@@ -5025,6 +5035,117 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+    // Compare the background validation chainstate's UTXO set hash against the hard-coded
+    // assumeutxo hash we expect.
+    //
+    // TODO: For belt-and-suspenders, we could cache an obfuscated version of the UTXO set
```
Removed.

jamesob commented on src/validation.cpp:
```diff
@@ -5025,6 +5035,117 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+
+    m_snapshot_validated = true;
+
+    // m_ibd_chainstate is the same as chainstate, per the assertion above.
+    // Modification happens to the argument (this symbol) to make it explicit that
+    // the caller might act specifically based on this value being different.
```
I must've been in a haze when I typed this - I have no idea what it means either. Removed.

jamesob commented on src/validation.cpp:
```diff
@@ -5128,3 +5130,28 @@ bool ChainstateManager::DetectSnapshotChainstate(CTxMemPool* mempool)
+    }
+
+    LogPrintf("[snapshot] tearing down snapshot datadir %s\n", fs::PathToString(*snapshot_datadir));
+    assert(fs::remove_all(*snapshot_datadir) > 0);
```
Both good points, both fixed. I'm now using `leveldb::DestroyDB()` to remove chainstate data.

jamesob commented on src/validation.cpp:
```diff
@@ -5174,3 +5323,44 @@ std::optional<int> ChainstateManager::getSnapshotHeight()
+    LogPrintf("[snapshot] deleting background chainstate directory (now unnecessary) (%s)\n",
+        fs::PathToString(old_chainstate));
+
+    // Instead of deleting the background chainstate directly, rename the old
+    // chainstate to chainstate_bak, rename the new chainstate into place, and then
+    // delete to avoid likelihood of corruption from mid-delete shutdown.
```
Removed comment

jamesob commented on src/validation.h:
```diff
@@ -772,11 +773,24 @@ class CChainState
+    MISSING_CHAINPARAMS,
+    STATS_FAILED,
+    HASH_MISMATCH,
+};
```
Done

jamesob reviewed (commented):
Thanks to everyone (esp. @ryanofsky) for the feedback so far.

I'm about to push the latest iteration of this branch. The one bit of feedback I'm not incorporating (yet) is the rename of the leveldb dir `chainstate_[blockhash]` to `chainstate_snapshot`, mostly because introducing a non-leveldb file (`blockhash` or whatever we eventually call it) complicates the `DestroyDB`/`removeSnapshotDatadir` process, since it's a non-leveldb file that leaves the directory dangling after the `leveldb::DestroyDB` call. But that said, I do want to go that route at some point.

Some notable changes resulting from the feedback so far:
- I've added utilities to create the coinsdb on disk (instead of in memory) within unittests; previously all chainstate data has been in memory for unittests, but this made testing snapshot cleanup basically impossible. I think having the default case be for coinsdb to be persisted to disk makes the tests more realistic anyway, which is good.
- Based upon Russ' good feedback, the unnecessary background validation chainstate data is now deleted at subsequent initialization (instead of at shutdown). It's plausible to me that we may be able to do this "inline" with snapshot validation, but I'll have to look into that and spend a few more days testing it.

jamesob committed:
docs: update assumeutxo.md

Include notes about the `chainstate_snapshot` rename as well as
updates for the included code.

jamesob committed:
validation: rename snapshot chainstate dir

This changes the snapshot's leveldb chainstate dir name from
`chainstate_[blockhash]` to `chainstate_snapshot`. This simplifies
later logic that loads snapshot data, and enforces the limitation
of a single snapshot at any given time.

Since we still need to persis the blockhash of the base block, we
write that out to a file (`chainstate_snapshot/base_blockhash`) for
later use during initialization, so that we can reinitialize the
snapshot chainstate.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>

jamesob committed:
init: add utxo snapshot detection

Add functionality for activating a snapshot-based chainstate if one is
detected on-disk.

Also cautiously initialize chainstate cache usages so that we don't
somehow blow past our cache allowances during initialization, then
rebalance at the end of init.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>

jamesob commented on src/validation.cpp:
```diff
@@ -5128,3 +5130,28 @@ bool ChainstateManager::DetectSnapshotChainstate(CTxMemPool* mempool)
+        // Chainstate isn't based on a snapshot.
+        return false;
+    }
+    std::optional<fs::path> snapshot_datadir = FindSnapshotChainstateDatadir();
```
I've simplified all this with the rename to `chainstate_snapshot`.

jamesob commented on src/validation.cpp:
```diff
@@ -5230,6 +5256,136 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+// because we need to avoid advancing the background validation chainstate
+// farther than the snapshot base block - and this function is also invoked
+// from within ConnectTip, i.e. from within ActivateBestChain, so cs_main is
+// held anyway.
```
Perhaps it isn't actually necessary to hold `cs_main` throughout this if I set `m_stop_use` immediately, and then ensure all chainstate accesses are going through `isUsable()` checks...

ryanofsky commented on src/validation.cpp:
```diff
@@ -5369,3 +5553,62 @@ std::optional<int> ChainstateManager::getSnapshotBaseHeight()
+    LogPrintf("[snapshot] deleting background chainstate directory (now unnecessary) (%s)\n",
+        fs::PathToString(ibd_chainstate_path));
+
+    fs::path tmp_old{ibd_chainstate_path + "_todelete"};
+
+    fs::rename(ibd_chainstate_path, tmp_old);
+    fs::rename(snapshot_chainstate_path, ibd_chainstate_path);
+
+    LogPrintf("[snapshot] moving snapshot chainstate (%s) to " /* Continued */
+        "default chainstate directory (%s)\n",
+        fs::PathToString(snapshot_chainstate_path), fs::PathToString(ibd_chainstate_path));
```
In commit "validation: add ChainMan logic for completing UTXO snapshot validation" (2efe7c7ec82e7885e07140e1deb5b5e9d6869f1d)

I'm suprised you can just rename directories while leveldb is still using them. If this works, I guess it's ok, but I think more ideally CChainstate objects would have `Destroy()` and `Rename(new_name)` methods and would do these operations in a safe way without ChainStateManager being involved at such a low level.

ryanofsky reviewed (commented):
This looks great! It is much simpler than I remember it previously.

I left some suggestions below (all minor, feel free to ignore them), and so far I reviewed all the non-test changes. I want to spend a little more time looking at the last commit, then review the tests, too.

- [X] 98fc81fbaaa02792b94264680cc0062ccfde9698 db: add StoragePath, Options to CDBWrapper/CCoinsViewDB (1/15)
- [X] f427976e6a3e024564fa824cc0578ee9b1285198 validation: add CChainState::destroyCoinsDB (2/15)
- [X] da4aaba0073108092f0dd0e02753eea253395c4e validation: rename snapshot chainstate dir (3/15)
- [X] 94a484340be246e38ba40644f4e2ba1913d22240 init: add utxo snapshot detection (4/15)
- [X] 5e8879c402b23f8c3b71287a1f4393caa73dfcbd validation: remove snapshot datadirs upon validation failure (5/15)
- [X] 89f3fc8c9205c34a28c5b892a979a9513cce0905 add ChainstateManager.getSnapshot{BaseHeight,BaseBlock}() (6/15)
- [X] f2abfde9f425c90bd6580ada35df94b35e130462 blockmanager: avoid undefined behavior during FlushBlockFile (7/15)
- [X] 70455311b695e65b847031a90fa870ab50519b48 validation: add CChainState::m_stop_use and ChainMan::isUsable (8/15)
- [X] c4774650050118c8659a636c2d826c3ee8a9b5bd add CChainState::HasCoinsViews() (9/15)
- [X] 2efe7c7ec82e7885e07140e1deb5b5e9d6869f1d validation: add ChainMan logic for completing UTXO snapshot validation (10/15)
- [ ] 223561af0854347c6e10ee81650cbeec7b0c0284 move-only: test: make snapshot chainstate setup reusable (11/15)
- [ ] c11d0ac32c5f150bfecd023fd455fbd916b93622 test: add reset_chainstate parameter for snapshot unittests (12/15)
- [ ] 26f6e02888f2cf3ea164c2780368d721736ec8c5 test: allow coinsdb to reside on-disk in tests (13/15)
- [ ] e02233a2b109ad4ac59a5a0cfbd83189cd2f4c61 test: add testcases for snapshot completion (14/15)
- [ ] 559b1ec7242f4636b03aa39b1ec1b629ff1414d7 docs: update assumeutxo.md (15/15)

ryanofsky reviewed (commented):
Few more review comments for the last code commit "validation: add ChainMan logic for completing UTXO snapshot validation" (2efe7c7ec82e7885e07140e1deb5b5e9d6869f1d)

Suggestion for PR description: Maybe add 2-3 sentences saying what assumeutxo functionality was already merged before this PR, and what changes are left to implement after this. It's easy to forget where these changes fit into the bigger picture, and helpful to be reminded!

jamesob commented on src/init.cpp:
```diff
@@ -1474,6 +1474,13 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+            case ChainstateLoadingError::SNAPSHOT_VALIDATION_COMPLETE:
+                // Retry LoadChainstate now that the fully-validated snapshot
+                // chainstate datadir has been moved to the `chainstate` directory.
+                continue;
```
Yeah, definitely agree. I'll take a look.

jamesob commented on src/dbwrapper.h:
```diff
@@ -212,6 +217,12 @@ class CDBWrapper
 
     std::vector<unsigned char> CreateObfuscateKey() const;
 
+    //! path to filesystem storage
+    const fs::path m_path;
+
+    //! whether or not the database resides in memory
+    bool m_is_memory;
```
While I agree, there are a few users of `CDBWrapper` outside the scope of this PR (`BaseIndex::DB`, `CBlockTreeDB`) - they seem to construct with non-empty paths even when being used in memory, so making this change will involve a few out-of-scope modifications. I agree with you that this is the right way to go, and would be a nice follow-on.

jamesob committed:
move-only-ish: init: factor out chainstate initialization

Moves chainstate initialization into its own function. This is necessary
to later support a more readable way of handling background-validation
chainstate cleanup during init, since the chainstate initialization
functions may need to be repeated after moving leveldb filesystem
content around.

This commit isn't strictly necessary, but the alternative is to (ab)use
the `while` loop in init.cpp with a `continue` on the basis of a
specific ChainstateLoadingError return value from LoadChainstate. Not
only is this harder to read, but it can't be unittested.

The approach here lets us consolidate background-validation cleanup to
LoadChainstate, and therefore exercise it within tests.

This commit is most easily reviewed with

  git diff --color-moved=dimmed-zebra --color-moved-ws=ignore-space-change

jamesob committed:
test: move-only-ish: factor out LoadVerifyActivateChainstate()

in TestingSetup(). This is used in the following commit to test
reinitializing chainstates after snapshot validation and cleanup.

Best reviewed with `git diff --color-moved=dimmed-zebra`.

jamesob committed:
add utilities for deleting on-disk leveldb data

Used in later commits to remove leveldb directories for
- invalid snapshot chainstates, and
- background-vaildation chainstates that have finished serving their
  purpose.

jamesob committed:
log: add LoadBlockIndex() message for assumedvalid blocks

I found this useful during unittest debugging.

jamesob reviewed (commented):
[`au.init.4`](https://github.com/jamesob/bitcoin/tree/au.init.4) -> [`au.init.6`](https://github.com/jamesob/bitcoin/tree/au.init.6)

[View range diff on GitHub](https://github.com/jamesob/bitcoin/compare/au.init.4..au.init.6)

Thanks so much for the continued and good review @ryanofsky. I've pushed (another) big update, but thankfully I think this PR is in better shape than ever.

Namely, we're now able to test the entire snapshot cleanup process from within the cpp tests, including filesystem operations and even a simulated node restart. This is a big improvement IMO, since we are now testing the full assumeutxo lifecycle from within cpp tests.

Other changes include
- removing unnecessary `m_snapshot_validated` state, 
- avoiding the confusing `continue`-based `init.cpp` control flow for reinitializing the new chainstate after snapshot cleanup - all that is contained within `LoadSnapshot()` now,
- catching all `fs::filesystem_error`s where applicable (and moving these calls out of asserts),
- adding the ability to store block tree db on-disk from cpp tests,
- removing the controversial `lowerCamelCase` method naming (see #24846),
- better assurance that leveldb objects are destructed before any fs operations (delete, rename)

I think this PR is in a good place, but I see that it needs another rebase so I'll start on that now...

jamesob commented on src/validation.cpp:
```diff
@@ -5230,6 +5256,136 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+      LogPrintf("[snapshot] something is wrong! validation chain " /* Continued */
+             "should not have continued past the snapshot origin\n");
+         handleInvalidSnapshot();
+         return {false, SnapshotCompletionError::IBD_TOO_FAR};
```
Fixed this to be clearer.

jamesob commented on src/validation.cpp:
```diff
@@ -5230,6 +5256,136 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+            "%s failed to validate the -assumeutxo snapshot state. "
+            "This indicates a hardware problem, or a bug in the software, or a "
+            "bad software modification that allowed an invalid snapshot to be "
+            "loaded. As a result of this, the node will shut down and move any "
```
Fixed.

ryanofsky reviewed (approved):
Code review ACK d988631b50b8b10a8dadae00fe61ebd32058a9d1. This is looking very good. The changes seem even more straightforward than before and the new tests are pretty clean and seem to cover the basic functionality well.

I think I'd still question the design decision (https://github.com/bitcoin/bitcoin/pull/24232#discussion_r861875528) not to delete the background chainstate after it's been validated until the next restart. IMO this makes things awkward in both the code and the tests because it adds an extra initialization side-quest that has to rehash the utxo set after it was already hashed, repeat initialization that was already done, and add hacky test code simulating a restart in order to reach this code path. I think if you just deleted the background chainstate and renamed the snapshot chainstate over it right after verifying the background chainstate, everything would be more straightforward. I don't see why it should be a problem because higher level code can already deal with switching to a new chainstate when the snapshot is loaded, so just unloading and reloading the chainstates without even changing which one is active seems like an easier problem. This does seems like something could be addressed in a followup, though.

ryanofsky commented:
> My only conceptual hangup at the moment is that we will need to do some kind of hacky DisconnectTippish thing of the snapshot base block to ensure that if the cleanup process isn't started or doesn't complete before shutdown, that we are able to catch it on subsequent boot.

I guess the last block on the background chain being connected is what triggers validating the background utxo set and deleting it, so you would want to leave the last block disconnected in order to trigger it again in case this is interrupted. This seems reasonable. Also seems reasonable that there could be some different trigger. I think the main improvement would be that if the backround validation and cleanup code is more integrated into normal block processing, init code can be simpler, and users wouldn't have a reason to stop and restart their nodes.

This seems reasonable. Also seems just as reasonable that there could be another that says if starting up and the background chain is fully downloaded, validate and delete it.

jamesob commented on src/validation.cpp:
```diff
@@ -5294,3 +5305,19 @@ ChainstateManager::~ChainstateManager()
+    if (!base_blockhash) {
+        return false;
+    }
+    this->InitializeChainstate(mempool, /*snapshot_blockhash=*/ *base_blockhash);
```
Fixed, thanks.

jamesob commented on src/node/chainstate.cpp:
```diff
@@ -9,6 +9,81 @@
+        }
+
+        // The on-disk coinsdb is now in a good state, create the cache
+        chainstate->InitCoinsCache(chainman.m_total_coinstip_cache * init_cache_fraction);
```
Fixed.

jamesob commented on src/validation.cpp:
```diff
@@ -5315,6 +5341,137 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+    assert(index_new.nHeight == snapshot_base_height);
+
+    auto handle_invalid_snapshot = [&]() EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
+        bilingual_str user_error = _(strprintf(
```
Fixed.

jamesob commented on src/validation.cpp:
```diff
@@ -5392,6 +5549,37 @@ bool ChainstateManager::DetectSnapshotChainstate(CTxMemPool* mempool)
+    AssertLockHeld(::cs_main);
+    // Should never be called on a non-snapshot chainstate.
+    assert(m_from_snapshot_blockhash);
+    fs::path snapshot_datadir = *FindSnapshotChainstateDatadir();
```
Fixed.

jamesob reviewed (commented):
Thanks for all the feedback there @ryanofsky. It took me much longer than I expected to address everything.

I am not going to address your recommendation here to swap out the chainstate data "inline" immediately after snapshot validation (vs. current approach of waiting until subsequent startup) because of the combination of two things:

- I'm running out of steam on assumeutxo. I have been working on this feature for 3 years and it isn't even existentially important to bitcoin. I have other stuff that is probably more valuable for me to work on and I'm pretty close to burnt out on this project.
- I don't see that approach as being substantially safer or more correct. In fact, making such a dramatic change as swapping out directories and reinitializing chainstates "in flight" seems fairly daring to me, vs. just halting use of one of the chainstates and doing the janitorial work at some point when the user is stopping/starting the bitcoind process.

I want to heartily thank you for your continued review of this PR, and I want to apologize for being so demoralized by the length of this project.

ariard reviewed (commented):
Reviewed until 2abc1b1, should review other commits soon.

> I'm running out of steam on assumeutxo. I have been working on this feature for 3 years and it isn't even existentially important to bitcoin. I have other stuff that is probably more valuable for me to work on and I'm pretty close to burnt out on this project

I still think assumeutxo is an interesting project to work on, especially to guarantee a smooth initial syncing process. If we aim to keep onboarding new full-nodes users in the coming decade to strengthen network decentralization it's more than nice to get assumeutxo in our feature set. It's also works well with a potential utreexo integration in core.

I would be interested to see a clear project status to know where to focus review time, neither the board or the meta-PR sounds to provide this information. I think it would also help to cut PRs in smaller chunks, even if there are not conceptually hard like this one. Hopefully assumeutxo should be close from a v0.1 release.

jamesob commented on src/node/utxo_snapshot.cpp:
```diff
@@ -0,0 +1,77 @@
+    afile << *snapshot_chainstate.m_from_snapshot_blockhash;
+
+    if (afile.fclose() != 0) {
+        LogPrintf("[snapshot] failed to close base blockhash file %s after writing\n",
```
Agree, maybe will do in a later change.

jamesob commented on src/validation.cpp:
```diff
@@ -5230,3 +5214,31 @@ ChainstateManager::~ChainstateManager()
+    assert(!m_snapshot_chainstate);
+    m_snapshot_chainstate = 
+        std::make_unique<CChainState>(mempool, m_blockman, *this, base_blockhash);
+    LogPrintf("Switching active chainstate to %s\n", m_snapshot_chainstate->ToString());
```
Done.

jamesob reviewed (commented):
Thanks for your review and encouragement, @ariard.

> I think it would also help to cut PRs in smaller chunks, even if there are not conceptually hard like this one.

In principle, I agree, and that's how I started the project. But I have found that even small PRs take months to merge (e.g. https://github.com/bitcoin/bitcoin/pull/23997 has been open for 6 months), and I have found that sadly the process encourages bundling large changesets together because otherwise progress feels impossibly slow. 

I have considered at various points doing the remainder of assumeutxo in a single PR because then at least review/test time could be consolidated on one changeset instead of a painfully lengthy series of smaller changes, each of which seems to take nearly half a year to get in.

jamesob commented:
> I would be interested to see a clear project status to know where to focus review time, neither the board or the meta-PR sounds to provide this information.

I have added an updated project status list on the parent PR (https://github.com/bitcoin/bitcoin/pull/15606#issue-421531882, cc @ariard). Please let me know if there's anything I can add that would be helpful. Only about half the subtasks are complete so far, but (I hope) we are farther along than halfway, as after this PR and #24008 are merged, most of the tricky logic and refactoring will be done.

jamesob commented on src/node/utxo_snapshot.cpp:
```diff
@@ -0,0 +1,77 @@
+    assert(snapshot_chainstate.m_from_snapshot_blockhash);
+
+    const std::optional<fs::path> chaindir = snapshot_chainstate.CoinsDB().StoragePath();
+    assert(chaindir); // Sanity check that chainstate isn't in-memory.
```
While I agree that "in-memory" is a little confusing in a vacuum, that term has had a very specific meaning for awhile in the context of chainstates and coinsdb - see e.g. the arguments `in_memory` on `CChainState::InitCoinsDB`, `CoinsViews::CoinsViews`, and `CCoinsViewDB::m_is_memory`.

ariard reviewed (commented):
Reviewed all the code changes until cbcc8bc.
Still have to check the correctness between the operations model laid out in `assumeutxo.md` and the code and the test coverage/changes.

_______

> In principle, I agree, and that's how I started the project. But I have found that even small PRs take months to merge (e.g. https://github.com/bitcoin/bitcoin/pull/23997 has been open for 6 months), and I have found that sadly the process encourages bundling large changesets together because otherwise progress feels impossibly slow. 

I understand that core review process might sounds slow and that large changesets are a way to keep the ball rolling forward from a PR author perspective. However, I think large changesets  might also produce the downside that they're are likely to discourage reviewers to get engaged, as the context for them to consider and evaluate is bigger. 

I would advise to split this PR in two chunks, one the logic for detecting and activating snapshot chainstates on start, the other one the logic to complete the snapshot validation process. Especially, the latter one, as it's modifying `ActivateBestChain()`, even superficially, it's one of the most critical path of the codebase and I think it's worthy of more scrutiny.

Though more generally, about the assumeutxo project itself, I think a majority of the contributors who partaked in the design of assumeutxo are not or less active anymore. Minding that the project is context-rich with changes in the validation engine, the caches, the disk I/O , the init sequence and the indexes, I think it's a pretty high bar to attract review. I would say assumeutxo would benefit from a renewed evangelism about why it matters and the upsides for the Bitcoin ecosystem as whole. I think it could be reasonable to give regular updates about the state of the project at meetings, the updated project status list sounds a good start.

Last, I believe a lot of opened PRs about validation and indexes could benefit from your review bandwidth. Speaking for myself, I always seek to have a positive review balance sheet between how much bandwidth I'm providing to the project and how much I'm requesting for my own PRs.

ryanofsky commented on src/validation.cpp:
```diff
@@ -5311,3 +5521,97 @@ std::optional<int> ChainstateManager::GetSnapshotBaseHeight()
+                  fs::PathToString(tmp_old));
+    }
+
+    LogPrintf("[snapshot] deleted background chainstate directory (%s)\n",
```
In commit "validation: add ChainMan logic for completing UTXO snapshot validation" (00b319a35cfb691a2fb831e84f63512412cae068)

Seems like this should be in an `} else {` block to avoid confusing `deletion of %s failed` message followed by `%s deleted` message

ryanofsky reviewed (approved):
Code review ACK d003388b07ed567a99f8c23f8a273145de6aa012. A lot of changes and cleanups since last review like simplifying the InitializeChainstate method. This looks very good and puts all the code needed to validate and clean up snapshots into place. I left more review comments, but they aren't important and can be ignored

jamesob reviewed (commented):
Rebased and addressed feedback. Thanks Russ and Antoine.

I'm approaching "thermonuclear levels of burnout" on this one, as the phrase goes. I'm getting bludgeoned by rebases, so any help reviewing this is appreciated. 

I've looked at splitting this out into two PRs and while it is possible it will not be trivial for me to do. If any potential reviewers feel strongly about doing that, I can.

jamesob commented:
Okay, I think it's pretty clear that the right move is to split this PR up into two separate ones (initialization vs. completion). I'll be opening those shortly.

ryanofsky reviewed (commented):
Code review ACK ac8c9d809214bd8dcdacdb66127cabf49788e097. Rebase and a lot of suggested cleanups since the last review, nothing major

I'm still a little confused about MaybeCompleteSnapshotValidation error handling, but in the worst case problems might just make debugging & error recovery less convenient so this is not a blocker

ariard commented on src/validation.h:
```diff
@@ -994,6 +988,13 @@ class ChainstateManager
+    //! snapshot that is in the process of being validated.
+    bool DetectSnapshotChainstate(CTxMemPool* mempool) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+
+    CChainState& ActivateSnapshotChainstate(CTxMemPool* mempool, uint256 base_blockhash) 
```
Well I don't have it anymore on the latest push of #25667 so all good.

ariard commented on src/node/utxo_snapshot.cpp:
```diff
@@ -0,0 +1,77 @@
+
+bool WriteSnapshotBaseBlockhash(CChainState& snapshot_chainstate)
+{
+    AssertLockHeld(::cs_main);
```
All good to me.

ariard commented on src/node/chainstate.cpp:
```diff
@@ -20,10 +20,26 @@ static std::optional<ChainstateLoadingError> CompleteChainstateInitialization(
+        return ChainstateLoadingError::ERROR_BAD_GENESIS_BLOCK;
+    }
+
+    // At this point blocktree args are consistent with what's on disk.
```
Right, that's accurate.

ariard commented on src/validation.cpp:
```diff
@@ -2750,6 +2750,14 @@ bool CChainState::ConnectTip(BlockValidationState& state, CBlockIndex* pindexNew
+    if (this != &m_chainman.ActiveChainstate()) {
+        // This call may set `m_stop_use`, which is referenced immediately afterwards in
+        // ActivateBestChain.
+        m_chainman.MaybeCompleteSnapshotValidation();
```
Good to me, it's a bit byzantine but that's allright.
